### PR TITLE
Adding support for ReciprocalEstimate and ReciprocalSqrtEstimate

### LINKED ISSUE #39971: Introduce Math/F.ReciprocalEstimate and Math/F.ReciprocalSqrtEstimate
## Background and Motivation

This would expose methods in `Math`/`MathF` which provide a method to calculate estimations for the reciprocal (`1 / x`), reciprocal square root (`1 / sqrt(x)`) and square root of a method. These are estimations, and are faster than their equivalent "exact" IEEE counterparts but are not necessarily deterministic on all systems.

These methods prove very useful in 3D graphics where you need to do these operations a lot and very fast, but exact accuracy is not key. On x64, they would map to `rcp` for `Reciprocal` and `rsqrt` for `ReciprocalSquareRoot`

## Proposed API

```diff
namespace System
{
    public static class Math
    {
+        public double ReciprocalEstimate(double x);
+        public double SqrtEstimate(double x);
+        public double ReciprocalSqrtEstimate(double x);
    }

    public static class MathF
    {
+        public float ReciprocalEstimate(float x);
+        public float SqrtEstimate(float x);
+        public float ReciprocalSqrtEstimate(float x);
    }
}
```
## Alternative Designs

`Approx` could be used instead of `Estimate`

## Outstanding Questions

Should similar methods be exposed on the various vector types?

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves #39971

### CODE REVIEW

**Comment by tannergooding** [comment]
`SqrtEstimate` was intentionally left off as neither x86 nor ARM provide an instruction for computing it and the proposed implementation of `x * ReciprocalSqrtEstimate(x)` returns `NaN` for `+/-0` and `+Infinity` which should be considered invalid behavior.

Checking for these two values will likely be slower than just calling `Sqrt` on modern hardware. On Skylake, we have approx. the following (https://uops.info):
* `sqrtss` is ~13 cycle latency and 6/3 (computed, measured) cycle throughput on port 0
* `rsqrtss` is ~5 cycle latency and 1/1 cycle throughput on port 0
* `mulss` is ~4 cycle latency and 0.5/0.5 cycle throughput on port 1
* `ucomiss` is ~3 cycle latency and 1/1 cycle throughput on port 0
* `cmpss` is ~4 cycle latency and 0.5/0.5 cycle throughput on port 1

Note: I haven't actually benchmarked the implementation that does `x * ReciprocalSqrtEstimate(x)` and checks for zero/+infinity, but given the general latencies, ports, and dependencies, I don't think we can make it faster.

File: src/libraries/System.Runtime.Extensions/tests/System/MathF.cs [hunk 1]
```diff
@@ -1421,6 +1421,84 @@ public static void Pow_IEEE(float x, float y, float expectedResult, float allowe
...
+        [InlineData( float.NaN,               float.NaN,              0.0f)]
+        [InlineData( 0.0f,                    float.PositiveInfinity, 0.0f)]
+        [InlineData( 0.318309886f,            3.14159274f,            CrossPlatformMachineEpsilon * 10000)] // value: (1 / pi)
+        [InlineData( 0.434294482f,            2.30258512f,            CrossPlatformMachineEpsilon * 10000)] // value: (log10(e))
+        [InlineData( 0.636619772f,            1.57079637f,            CrossPlatformMachineEpsilon * 10000)] // value: (2 / pi)
+        [InlineData( 0.693147181f,            1.44269502f,            CrossPlatformMachineEpsilon * 10000)] // value: (ln(2))
+        [InlineData( 0.707106781f,            1.41421354f,            CrossPlatformMachineEpsilon * 10000)] // value: (1 / sqrt(2))
+        [InlineData( 0.785398163f,            1.27323949f,            CrossPlatformMachineEpsilon * 10000)] // value: (pi / 4)
+        [InlineData( 1.0f,                    1.0f,                   CrossPlatformMachineEpsilon * 10000)]
+        [InlineData( 1.12837917f,             0.886226892f,           CrossPlatformMachineEpsilon * 1000)]  // value: (2 / sqrt(pi))
```
Comment (tannergooding) [inline src/libraries/System.Runtime.Extensions/tests/System/MathF.cs:1450]: This was the "worst" case of difference that was measured. 

```
1 / 1.12837917f                  == 0.886226892f
ReciprocalEstimate(1.12837917f ) == 0.886352539f
```

This gives a difference of ~`0.000125647` in expected vs actual. The instruction documents a maximum error of `1.5 * 2^-12` which is `0.0003662109375` while scaling the `CrossPlatformMachineEpsilon` allows for an error of up to `0.000476837158`

Comment (pgovind) [reply src/libraries/System.Runtime.Extensions/tests/System/MathF.cs:1450]: Just for myself, where is this instruction? In the Intel/AMD manuals?

**Review by pgovind — APPROVED** [review]
Very clean. LGTM!

File: src/libraries/System.Private.CoreLib/src/System/Math.cs [hunk 2]
```diff
@@ -950,6 +950,36 @@ public static double MinMagnitude(double x, double y)
             return y;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (pgovind) [inline src/libraries/System.Private.CoreLib/src/System/Math.cs:953]: Add docs?

File: src/libraries/System.Runtime.Extensions/tests/System/MathF.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Runtime.Extensions/tests/System/MathF.cs:1450]: `rcpss` and `rsqrtss`, they would be in Volume 2 of the Intel manuals (https://software.intel.com/content/www/us/en/develop/articles/intel-sdm.html) and Volume 4 of the AMD manuals (https://developer.amd.com/resources/developer-guides-manuals/)

File: src/libraries/System.Private.CoreLib/src/System/Math.cs (hunk 2 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Math.cs:953]: Yes, need to do that for this one.

Actually, @carlossanlop is the bot running, this seems to be missing that tag reminding me to add them 😄

Comment (pgovind) [reply src/libraries/System.Private.CoreLib/src/System/Math.cs:953]: A bot added the "new-api-needs-doc" label already right? Or is there another bot?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Math.cs:953]: Ah it's, there, I just missed it completely. It blends in with dark theme:
image

**Comment by danmosemsft** [comment]
Needs a bit more epsilon on musl eg
```
Assert.Equal() Failure\nExpected: -3.14159274\nActual:   -3.1484375


Stack trace
   at System.Tests.MathFTests.AssertEqual(Single expected, Single actual, Single variance) in /_/src/libraries/System.Runtime.Extensions/tests/System/MathF.cs:line 137
   at System.Tests.MathFTests.ReciprocalEstimate(Single value, Single expectedResult, Single allowedVariance) in /_/src/libraries/System.Runtime.Extensions/tests/System/MathF.cs:line 1460
```

**Comment by tannergooding** [comment]
> Needs a bit more epsilon on musl eg

Yep, it turns out a single `Newton-Raphson` iteration on ARM64 can be less precise than `1.5 * 2^-12`. In this case it was somewhere between `1.5 * 2^-6` and `1.5 * 2^-7`

**Pull request merged**